src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets

```diff
@@ -92,6 +92,23 @@ Copyright (c) .NET Foundation. All rights reserved.
     </_GenerateRestoreGraphProjectEntryInputProperties>
   </PropertyGroup>
 
+  <ItemDefinitionGroup Condition="'$(ManagePackageVersionsCentrally)' == 'true' And '$(RestoreEnableGlobalPackageReference)' != 'false'">
```

Comment:
This should be in `NuGet.props` if possible. The reason is that if others were to override or update existing definition, then the last one defined wins which is now this but this should be user overridable! Hence the location change!!

Reply:
In order for GlobalPackageReference to work as intended, we'll strongly recommend people use it with the default functionality.  Other parts of the toolset like Visual Studio might depend on certain things and so user customizations are not guaranteed to work end-to-end.  The functionality is configured in such a way that you can enable/disable assets selection an escape hatch for potential corner cases.  You can also of course do a `<GlobalPackageReference Update="" />` later on to completely customize how this functions.  But again, I really don't want people to customize it too much and get themselves into an unsupported state.

Reply:
This doesn't affect Visual Studio tooling unless we specifically refer to these defaults hard-coded in the project system which is highly unlikely since package reference concept already exist in VS. This move should be safe.

Reply:
Item Definitions should always be imported before item declarations themselves, though this doesn't matter with MSBuild evaluation but this also simplifies understanding.

Reply:
> This doesn't affect Visual Studio tooling unless we specifically refer to these defaults hard-coded in the project system which is highly unlikely since package reference concept already exist in VS

We'd want to give ourselves the option to be able to do that. 

> now this but this should be user overridable

This whole feature is syntactic sugar. Overriding it and tweaking stuff on top it seems like an overkill to me.

Reply:
I see `'$(RestoreEnableGlobalPackageReference)' != 'false`

Is the idea that the feature is enabled by default? 
Should the prop name instead be `RestoreDisable...` and check for true instead?

Reply:
> Overriding it and tweaking stuff on top seems like overkill to me.

In the current implementation, I could just disable and create a new definition but it also disables the actual functionality. That's why `RestoreEnableGlobal~~~Assets~~` properties are there but as you said per <https://github.com/NuGet/NuGet.Client/pull/4680#discussion_r919585461> that's overkill to disable individual metadata.

That's why I suggested placing the definition in `NuGet.props` and remove those metadata opt-outs!

Reply:
I don't think you need to have it in the props to avoid the opt-outs. They can just be removed. 

> In the current implementation, I could just disable and create a new definition but it also disables the actual functionality

MSBuild is extremely powerful.
If you disable the functionality and reimplement it on your own, that already becomes an unsupported scenario.

As I mentioned earlier, this is a helper/syntactic sugar addition. 
If you want to implement your customer functionality, you can always just use a different item name and use that throughout your repo.

Reply:
That's not what I meant. As you and @zivkan said opt outs are overkill. I agree. We can remove them. But we lose the customization that @jeffkl proposes. What the proposal says that by default global packages are only build-time dependencies but can be a toolset or compile-time dependencies as long as we can change the shape of the item definition. For that to be possible, it needs to be placed in props.

With what I read about writing MSBuild targets/tasks, default properties and item definitions (_not declarations_) should be in props unless we only have targets in which case it should be first in the targets file.

Also, only people who have atleast some insight into MSBuild would know about item definitions, so people accidentally messing up, won't be the case here.

Reply:
> What the proposal says that by default global packages are only build-time dependencies but can be a toolset or compile-time dependencies as long as we can change the shape of the item definition. For that to be possible, it needs to be placed in props.


If you want to have a package that uses compile, you can simply define a `PackageReference`. 
I don't see why you need a separate item from that. 
You can place the PackageReference in any file in your repo.

Reply:
I'm aware of that but this is just a clean solution. If I introduce item definition for `PackageReference` it messes up for all package references. If I introduce a new item, I have to manually add it to `PackageReference` item, which doesn't work with central versioning. This "global" item definition and the implementation itself solves most of the issues I face but the only thing is, for some projects, I have compile-time global dependencies which needed to be included/excluded dynamically.

Thus, I want to be able to say, for these projects, override the built-in definition to use mine with my default metadata. This would only work if the item definition were to be placed in props!

Reply:
I'm not sure I follow. 
You can add an implementation for another item in the exact same way in a DIrectory.Build.props/targets and it'd just work. 

Overriding the build definition is imo the design for this item. It's an optimization. If people want to have their own custom things, then so be it.

Not everything has to be built into NuGet. Maintainability is a concern for a tool like NuGet has so many little features

Reply:
Adding another item means that I need to add the exact same implementation that is added here in the exact same way. That's too much since I only want to change the default metadata for all GPR items.

I could do `Update` but that might work only for items that were added before the update statement and it also overrides any custom metadata that is set. So, overriding the item definition is our only option.

You seem to strongly object to this. I'm not seeing any issue with moving the definition to props. Is there anything I'm missing?

Reply:
A very important for UX, is when things `work as expected`. 
If someone non-Nirmal4G sees a GlobalPackageReference declared, they have expectations about what that reference does. 

> Adding another item means that I need to add the exact same implementation that is added here in the exact same way. That's too much since I only want to change the default metadata for all GPR items.

It's a 5 line change, I don't necessarily see this as something significantly difficult. 
Adding things in the tooling itself means something that'd have to be maintained for years to come. 

> You seem to strongly object to this. I'm not seeing any issue with moving the definition to props. Is there anything I'm missing?

This is syntactic sugar that would "just work" as designed in a large majority of cases. 
My reasoning is 2-fold: 
* "Works as expected"
* Maintenance. All functionality added to the product is something that'd be maintained for years to come.

Reply:
I see. Then again, your conditions would still be satisfied, if we moved it. But I see the reasoning to avoid surprise bug reports.

Anyway, I figured that this can be overriden in directory targets since NuGet targets are imported before them. Previously I thought it is being imported after directory targets, that's why I suggested the move!

***IMO, default properties and item definitions should be imported before everything else.***

Reply:
> IMO, default properties and item definitions should be imported before everything else.

Personally, I'm comfortable with whatever @jeffkl decides on that.